### PR TITLE
Bug 1160399 and the rest: rework how we handle DB stuff.

### PR DESCRIPTION
**Pull request by rnewman**
(I reviewed and cherry-picked Wes's patch that's in this branch.)

Naturally this doesn't build at most intermediate stages, but the individual commits should make sense.

Still to come: rewriting the rest of the disabled tests.

### CODE REVIEW

**Comment by ncalexan** [comment]
This is one heroic patch series.

I cannot guarantee we have not regressed things -- possibly in serious ways -- because of the scope of the patch; the fact that it's very hard to compare before and after when the first patch is to remove everything; and because I am not intimately familiar with the original code.

However, the final result reads better than what I was familiar with -- even the hairy multi-level branching database insertions.

SHIP IT!

**Comment by rnewman** [comment]
Danke. The original branch in `rnewman/db` doesn't delete anything until much later; I kept that around to show some side-by-side.

I'm going to get some more of the few disabled tests back up and running before landing. Thanks for the thorough review!

File: Client/Frontend/Browser/PasswordHelper.swift [hunk 1]
```diff
@@ -88,14 +88,15 @@ class PasswordHelper: BrowserHelper {
 
         profile.passwords.get(QueryOptions(filter: password.hostname), complete: { data in
             for i in 0..<data.count {
-                let savedPassword = data[i] as! Password
-                if savedPassword.username == password.username {
-                    if savedPassword.password == password.password {
+                if let savedPassword = data[i] {
```
Comment (wesj) [inline Client/Frontend/Browser/PasswordHelper.swift]: We should probably make this a "for password in data", but I remember hitting some swift errors originally when i implemented that maybe it doesn't work. Still data[i] should not be nil here. If it is, something is busted.

File: Client/Frontend/Home/BookmarksPanel.swift [hunk 2]
```diff
@@ -11,7 +11,12 @@ class BookmarksPanel: SiteTableViewController, HomePanel {
 
     override var profile: Profile! {
         didSet {
-            profile.bookmarks.modelForRoot(self.onNewModel, failure: self.onModelFailure)
+            // Until we have something useful to show for desktop bookmarks,
+            // only show mobile bookmarks.
+            // Note that we also need to build a similar kind of virtual hierarchy
+            // to what we have on Android.
+            profile.bookmarks.modelForFolder(BookmarkRoots.MobileFolderGUID, success: self.onNewModel, failure: self.onModelFailure)
```
Comment (wesj) [inline Client/Frontend/Home/BookmarksPanel.swift]: Guids should not be exposed anywhere... maybe that's impossible to fix.

File: Client/Frontend/Home/HistoryPanel.swift [hunk 3]
```diff
@@ -25,9 +25,7 @@ class HistoryPanel: SiteTableViewController, HomePanel {
     private var sectionOffsets = [Int: Int]()
 
     override func reloadData() {
-        let opts = QueryOptions()
-        opts.sort = .LastVisit
-        profile.history.get(opts).uponQueue(dispatch_get_main_queue()) { result in
+        profile.history.getSitesByLastVisit(100).uponQueue(dispatch_get_main_queue()) { result in
```
Comment (wesj) [inline Client/Frontend/Home/HistoryPanel.swift]: Can you file a bug to make history dynamically fetch more data as you scroll? I don't really mind a limit on TopSites, but having one on History feels kinda bad to me.

Comment (wesj) [inline Client/Frontend/Home/HistoryPanel.swift]: Not to say I can't live with it for now. :)

File: Client/Frontend/Home/BookmarksPanel.swift (hunk 2 above)
Comment (rnewman) [inline Client/Frontend/Home/BookmarksPanel.swift]: I guess we could add `modelForMobileBookmarksFolder` or something… care enough for me to do that?

File: Client/Frontend/Home/HistoryPanel.swift (hunk 3 above)
Comment (rnewman) [inline Client/Frontend/Home/HistoryPanel.swift]: Yeah, good idea.

Comment (rnewman) [inline Client/Frontend/Home/HistoryPanel.swift]: https://bugzilla.mozilla.org/show_bug.cgi?id=1162742

**Comment by rnewman** [comment]
I'm rebasing this for landing here:

https://github.com/mozilla/firefox-ios/tree/rnewman/db-rebase-landing

File: Client/Frontend/Home/BookmarksPanel.swift (hunk 2 above)
Comment (wesj) [inline Client/Frontend/Home/BookmarksPanel.swift]: I think I'd like to move all of the guid/folders to an enum, but i'm fine with lots of small follow ups for non user facing stuff.

**Comment by rnewman** [comment]
Landed: https://github.com/mozilla/firefox-ios/commit/6db9f8b2a2b9f5cd7b74a941f836c14318d378f6

**Pull request closed without merging**